fml/message_loop_impl.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop_impl.h"
+#include "lib/ftl/build_config.h"
+
+#if OS_MACOSX
+
+#include "flutter/fml/platform/darwin/message_loop_darwin.h"
+using PlatformMessageLoopImpl = fml::MessageLoopDarwin;
+
+#else
+
+#error This platform does not have a message loop implementation.
+
+#endif
+
+namespace fml {
+
+ftl::RefPtr<MessageLoopImpl> MessageLoopImpl::Create() {
+  return ftl::MakeRefCounted<::PlatformMessageLoopImpl>();
+}
+
+MessageLoopImpl::MessageLoopImpl() : order_(0) {}
+
+MessageLoopImpl::~MessageLoopImpl() = default;
+
+void MessageLoopImpl::PostTask(ftl::Closure task, ftl::TimePoint target_time) {
+  if (task == nullptr) {
+    return;
+  }
+
+  WakeUp(RegisterTaskAndGetNextWake(task, target_time));
+}
+
+void MessageLoopImpl::RunExpiredTasksNow() {
+  WakeUp(RunExpiredTasksAndGetNextWake());
+}
+
+void MessageLoopImpl::SetTaskObserver(MessageLoop::TaskObserver observer) {
+  ftl::MutexLocker lock(&task_observer_mutex_);
+  task_observer_ = observer;
+}
+
+ftl::TimePoint MessageLoopImpl::RegisterTaskAndGetNextWake(
+    ftl::Closure task,
+    ftl::TimePoint target_time) {
+  FTL_DCHECK(task != nullptr);
+  ftl::MutexLocker lock(&delayed_tasks_mutex_);
+  delayed_tasks_.push({++order_, std::move(task), target_time});
+  return delayed_tasks_.top().target_time;
+}
+
+ftl::TimePoint MessageLoopImpl::RunExpiredTasksAndGetNextWake() {
+  std::vector<ftl::Closure> invocations;
+
+  {
+    ftl::MutexLocker lock(&delayed_tasks_mutex_);
+
+    if (delayed_tasks_.empty()) {
+      return ftl::TimePoint::Max();
+    }
+
+    auto now = ftl::TimePoint::Now();
+    while (!delayed_tasks_.empty()) {
```

Comment:
I don't think I completely understand this comment. Does the vector contain the list of all pending task invocations? In that case, I can't simply swap the vector out because when I later dump it into the priority queue to figure out invocation order, I may not end up actually having to execute all tasks in the queue (if the `target_time` > `now`). Now, even though I have new tasks piling up in the vector just fine, I have a few leftover tasks from the priority queue step that I would need to add back to this vector (and would need to re-acquire the mutex for). Which is fine I guess. Let me know if I understood you correctly. Will update.